video/out/gpu/context.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "video/out/vo.h"
+
+#include "config.h"
+#include "ra.h"
+
+enum ra_api_type {
+    RA_API_ANY    = 0,
+    RA_API_OPENGL,
+};
+
+struct ra_ctx_opts {
+    int allow_sw;      // allow software renderers
+    int want_alpha;    // create an alpha framebuffer if possible
+    int debug;         // enable debugging layers/callbacks etc.
+    bool probing;      // the backend was auto-probed
+    int swchain_depth; // max number of images to render ahead
+};
+
+struct ra_ctx {
+    struct vo *vo;
+    struct ra *ra;
+    struct mpv_global *global;
+    struct mp_log *log;
+
+    struct ra_ctx_opts opts;
+    const struct ra_ctx_fns *fns;
+
+    bool flip_v; // flip the rendered image vertically (set by the ra_ctx)
+    void *priv;
+
+#if HAVE_GL
+    struct ra_gl_ctx *gl_priv;
+#endif
+};
+
+// The functions that make up a ra_ctx.
+struct ra_ctx_fns {
+    enum ra_api_type type; // API type (for --gpu-api)
+    const char *name;      // name (for --gpu-context)
+
+    // Resize the window, or create a new window if there isn't one yet.
+    // Currently, there is an unfortunate interaction with ctx->vo, and
+    // display size etc. are determined by it.
+    bool (*reconfig)(struct ra_ctx *ctx);
+
+    // Gets the current framebuffer depth in bits (0 if unknown). Optional.
+    int (*depth)(struct ra_ctx *ctx);
+
+    // Retrieves a screenshot of the framebuffer. These are always the right
+    // side up. Optional.
+    struct mp_image *(*screenshot)(struct ra_ctx *ctx);
+
+    // Called when rendering starts. Returns NULL on failure. This must be
+    // followed by present_frame, to submit the rendered frame.
+    struct ra_tex *(*start_frame)(struct ra_ctx *ctx);
+
+    // Present the frame. Issued in lockstep with start_frame, with rendering
+    // commands in between. The `frame` is just there for timing data, for
+    // contexts smart enough to do something with it.
+    bool (*submit_frame)(struct ra_ctx *ctx, const struct vo_frame *frame);
+
+    // Performs a buffer swap. This blocks for as long as necessary to meet
+    // params.swchain_depth, or until the next vblank (for vsynced contexts)
+    void (*swap_buffers)(struct ra_ctx *ctx);
```

Comment:
I don't get the difference to submit_frame?

Did you think about advanced stuff like async/scheduled present and presentation feedback?

Reply:
The simple answer is that `submit_frame` is called from `draw_frame` and `swap_buffers` is called from `flip_frame`. In cases like vulkan, and opengl with `opengl-early-flush`, we want to submit the work as early as possible (and flush it to the GPU) and only block on the swapbuffers “later”.